cmd/otel-allocator/allocation/consistent_hashing.go

```diff
@@ -0,0 +1,189 @@
+package allocation
+
+import (
+	"fmt"
+	"github.com/buraksezer/consistent"
+	"github.com/cespare/xxhash/v2"
+	"github.com/go-logr/logr"
+	"github.com/open-telemetry/opentelemetry-operator/cmd/otel-allocator/diff"
+	"github.com/prometheus/client_golang/prometheus"
+	"net/url"
+	"sync"
+)
+
+var _ Allocator = &consistentHashingAllocator{}
+
+const consistentHashingStrategyName = "consistent-hashing"
+
+type hasher struct{}
+
+func (h hasher) Sum64(data []byte) uint64 {
+	return xxhash.Sum64(data)
+}
+
+type consistentHashingAllocator struct {
+	// m protects Consistent for concurrent use.
+	m sync.RWMutex
+
+	hasher *consistent.Consistent
+
+	// collectors is a map from a Collector's name to a Collector instance
+	collectors map[string]*Collector
+
+	// targetItems is a map from a target item's hash to the target items allocated state
+	targetItems map[string]*TargetItem
+
+	log logr.Logger
+}
+
+func newConsistentHashingAllocator(log logr.Logger) Allocator {
+	config := consistent.Config{
+		PartitionCount:    1061,
+		ReplicationFactor: 5,
+		Load:              1.1,
+		Hasher:            hasher{},
+	}
+	consistentHasher := consistent.New(nil, config)
+	return &consistentHashingAllocator{
+		hasher:      consistentHasher,
+		collectors:  make(map[string]*Collector),
+		targetItems: make(map[string]*TargetItem),
+		log:         log,
+	}
+}
+
+// addTargetToTargetItems assigns a target to the collector based on its hash and adds it to the allocator's targetItems
+// This method is called from within SetTargets and SetCollectors, which acquire the needed lock.
+// This is only called after the collectors are cleared or when a new target has been found in the tempTargetMap
+// INVARIANT: c.collectors must have at least 1 collector set
+func (c *consistentHashingAllocator) addTargetToTargetItems(target *TargetItem) {
+	colOwner := c.hasher.LocateKey([]byte(target.Hash()))
+	targetItem := &TargetItem{
+		JobName:       target.JobName,
+		Link:          LinkJSON{Link: fmt.Sprintf("/jobs/%s/targets", url.QueryEscape(target.JobName))},
+		TargetURL:     target.TargetURL,
+		Label:         target.Label,
+		CollectorName: colOwner.String(),
+	}
+	c.targetItems[targetItem.Hash()] = targetItem
+	c.collectors[colOwner.String()].NumTargets++
```

Comment:
Because we might be reallocating all targets, we should check if the targetItem already exists and, if so, decrement its collector's `NumTargets` count (and set that metric too).

Reply:
Ah good point! We may be getting occasionally saved currently by the fact we set it equal to the numTargets so as we assign it with least weighted we end up just setting everything correctly. This is a good call out though, ultimately this metric should be done with an async gauge rather than calling it on every method call. lemme think about the best way to fix this one